Issues/93

bsb20 opened the pull request:
Support for async extension hat blocks, as discussed with @cwillisf .  Extension hat blocks in the offline editor can now properly handle their "synchronous" javascript calls happening asynchronously.  This fixes  [93](https://github.com/LLK/scratch-flash-offline/issues/93).

Commit:
Handle extension name containing '.'

Commit:
Add WeDo 2.0 extension to the list in ExtensionManager

Commit:
Load extensions using resetPlugin callback

This allows `resetPlugin()` to finish completely before
loading extensions, fixing a bug in the offline editor.

Commit:
Fix warnings

Commit:
Update WeDo icons for transparent background

Resolves LLK/scratch-device-manager#12

Commit:
Rename original WeDo extension back to 'LEGO WeDo'

Changing the internal name would break projects which contain a
reference to this extension. If we really want it to show up as 'LEGO
WeDo 1.0' in the extension library we should add a way to separate the
display name of an extension from its internal name.

Commit:
Added support for extension display name

    -Can now specify displayName which can differ from name
    -displayName defaults to name if not specified

Commit:
Reset requester state when args might have changed

The `collectArgs()` function runs when anything changes in the arguments
of a block or any of its descendants, recursively. If this is called on
a Requester block which currently has a value, reset / discard that
value in case it's no longer valid given the argument change.
This resolves LLK/scratchx#61

Commit:
Added support for async hat blocks

    -Specified by 'H' in extensions
    -Will wait for callback to be called providing a predicate name
    -When called, will make call to evaluate the named predicate with
    most recent args

Commit:
Added support for async hat blocks

    -Specified by 'H' in extensions
    -Will wait for callback to be called providing a predicate name
    -When called, will make call to evaluate the named predicate with
    most recent args

Commit:
Fixed handling of async hat responses

    -Be ready to deal with 'false' in case the extension status is not
    'ready'

Commit:
Fixed handling of async hat responses

    -Be ready to deal with 'false' in case the extension status is not
    'ready'

Commit:
Adding 'forcedRequester' hat blocks for offline

Commit:
Adding 'forcedRequester' hat blocks for offline

Commit:
Changed async hat blocks for offline extensions

    -This may have broken 'H' extension hat block behavior
    -This should be the model used, thought, and the normal async hat
    block pattern should change to accomodate it
    -Offline extension hats are now marked 'hH', are 'forcedRequesters',
    and are handled similiarly to async hat blocks in the ScratchRuntime

Commit:
Changes to new async hat blocks

    -True async hat blocks should not wait to refire

Commit:
Added valid XML check to SVG decoding

    -As a catch all, you must be able to make a valid XML object from
    the (potentially) SVG data
    -Adding this to isSVGData seems to engage the preexisting error
    handling functionality

Commit:
Move 'when Stage clicked' to blocks strings

The string "when Stage clicked" was showing up in the "UI" strings
export (the "Scratch Editor" project in Pootle) instead of the
"commands" export (the "Scratch Blocks" project in Pootle). This change
fixes that, resolving #1031.

Commit:
Poor type safety in endCurrentTool caused err. 1096

    -This is not a permanent fix, but solves the problem while touching
    at little as possible, and will not throw even if unexpected types
    are received.

Commit:
added check to endCurrentTool

    -Non-selection objects will be wrapped in a Selection
    -Selection items must have a parent, so we should check for this

Commit:
Merge pull request #1037 from bsb20/hotfix/clone-tool

Poor type safety in endCurrentTool caused err. 1096

Commit:
Bump version to v444.1

cwillisf committed:
Merge pull request #1009 from LLK/wedo2

Add the new WeDo 2.0 extension

Commit:
v444.2 version bump

Commit:
Fix "About" URL for WeDo 2.0 extension

Commit:
v444.3 version bump

Commit:
Use U+001F to separate extension name from op

Prior to the WeDo 2.0 extension, Scratch assumed that an extension's
name would never contain `.` -- Scratch would split an extension op into
an extension name and an op within that extension by looking for the
first `.`.

With the WeDo 2.0 extension, I changed the code to split on the last
`.`. Unfortunately this broke extensions which use `.` in their blocks.

This change uses U+001F (Unicode "Unit Separator") as the separator
instead of `.` since U+001F is very unlikely to be contained in an
extension's name or blocks. The separator and unpacking logic have been
centralized in `ExtensionManager` as well.

This resolves #1044

Commit:
Use first extension separator, not last

Commit:
Initial guess at new media lib dirs

Commit:
Changed note picker behavior

    -Used to play note for 3 seconds
    -Now runs block.demo which is like clicking the block
    -The effect is, the note plays for the specified length

Commit:
v444.4 version bump

Commit:
Made demo block not visible

    -Made running as a 'background' thread unnecessary

Commit:
Merge pull request #1049 from bsb20/issues/1048

Changed note picker behavior

cpseager commented on src/extensions/ExtensionManager.as:
```diff
@@ -212,6 +212,10 @@ public class ExtensionManager {
+				if(spec[0] == 'h') {
+					// same as above, but force to use async hat blocks
+					spec[0] = 'hH';
+				}
```
@bsb20 This is duplicating the existing issue with rR in a saved .sb2 file not reverting to r when opened online. In usability this means the checkbox (e.g. beside tilt or distance in WeDo) is permanently lost to display the watcher on the stage (even when the sb2 file is opened online). 

Basically a saved .sb2 file can end up containing either 'r' or 'rR' and therefore this check should work both ways - i.e. r to rR if offline and rR to r when online. Same for h and hH.

cwillisf replied:
@cpseager @bsb20 I think that's an argument to either:
- make sure that `'rR'` gets saved as `'r'` in an SB2, or
- use another way to decide to use this behavior instead (obviating the need for the `'r'` -> `'rR'` swap in the first place).

I don't think I was aware of this problem. I've filed a new issue to track it, since I didn't find one in the editor repositories: https://github.com/LLK/scratch-flash/issues/1051

cpseager replied:
I can't find the issue either, but the attempted fix was previously noted in the (now obsolete) #756

cwillisf replied:
I think I've fixed this in #1053. If we're all happy with #1053 then this PR may need some adjustment as well, but I don't think it'll take too much.

cwillisf commented on src/extensions/ExtensionManager.as:
```diff
@@ -212,6 +212,10 @@ public class ExtensionManager {
 					// 'r' is reporter, 'R' is requester, and 'rR' is a reporter forced to act as a requester.
 					spec[0] = 'rR';
 				}
+				if(spec[0] == 'h') {
```
This line could use an `else`

cwillisf committed:
Merge pull request #1045 from LLK/fix-1044-extension-separator

Fix 1044 extension separator

cwillisf committed:
Merge pull request #1025 from LLK/fix-scratchx-61

Reset requester state when args might have changed

Commit:
Eliminate 'rR' block type hack

In the offline editor, reporters must be treated as requesters. This had
been implemented by converting `'r'` blocks into `'rR'` blocks, but this
ended up "leaking" into SB2s saved from the offline editor, which in
turn caused problems with the online editor.

Now, the `Block` constructor calls the new `shouldForceAsync()` method
on `ExtensionManager` to determine whether or not to set `forceAsync`
(renamed from `forceRequester`). The block type/spec is never altered.

In support of this change, the logic for showing the watcher checkbox
was updated to base its decision on the constructed block rather than
the block spec.

Commit:
Merge pull request #1033 from bsb20/issues/246

Added valid XML check to SVG decoding

cwillisf commented:
:+1: overall, pending changes related to #1053
@bsb20, let me know if you'd like to discuss this or my changes in more detail :)

cwillisf committed:
Merge pull request #1038 from LLK/translation-fixes-1031

Move 'when Stage clicked' to blocks strings

Commit:
removed new categories from backdrops

Commit:
Merge pull request #1055 from bsb20/issues/1015

Issues/1015

Commit:
Merge pull request #1053 from LLK/fix-1051

Eliminate 'rR' block type hack

Commit:
Changes pursuant to 1053

    -Forced async hat blocks need to use the same mechanism
    -Will test in offline editor

Commit:
Fixing missed forcedRequester removal

bsb20 commented:
Tested with the offline editor and the prototype async hat blocks in the online editor, looks good.

cwillisf commented on src/extensions/ExtensionManager.as:
```diff
@@ -342,18 +371,53 @@ public class ExtensionManager {
+						app.externalCall('ScratchExtensions.getReporter', function(v:*):void {
+							value = v;
+						}, ext.name, responseObj.predicate, args);
+					}
```
If `responseObj` has no `predicate` property here, is that a violation of the async hat block API?
If so, there should probably be an `else` here which logs a warning to the JS console to help out the extension author.
If not, does this need an `else` which sets `value` to... something? `true` maybe?

bsb20 replied:
Good point, fixed in [this commit](https://github.com/LLK/scratch-flash/pull/1032/commits/0af72a727824aa20d373071f0bb5defd66728834)

Commit:
If no predicate is specified, fire async hat

    -This mean if I'm not interested in evaluating my args again,
    and I just want to fire on every event, I can do so
    -In this case, saves on the latency of a roundtrip to/from js

cwillisf commented:
:+1:
